README.md
# Bamboo

Bamboo is a high-level api for ElasticSearch that uses a pandas-style interface for querying and aggregations. It was created to provide data scientists and other developers a means of interacting with ElasticSearch without having to learn its cumbersome query language.

DataFrame attributes are created dynamically from field names at initialization time. This allows the team in charge of updating an index or its mapping to be separate from the team using the index for analytics. This can be a significant boon for indices with frequently changing mapping definitions.

## Features

-  Introspection of fields and types
-  Autocomplete for field names
-  Autocomplete for method-based operations based on field type
-  Common syntactic operators regardless of field type
-  Queries are built dynamically and only executed when data is requested

## Installation

```shell
$ pip install elasticsearch-bamboo
...
```

## Example use

Simple use-case:

```python
>>> from bamboo import DataFrame

# by default we connect to localhost:9200
>>> df = DataFrame(index='my_index')

>>> df = df[df.age > 21]
>>> df.count()
42
>>> df.take(10)
[{...}]
```

## Usage

### Configuration

Connection parameters for ElasticSearch and Bamboo can be set several ways.

#### Helper function

The helper function only needs to be called once. After that, its state will persist throughout your session. It accepts the same parameters as the low-level ElasticSearch api. If no config is declared then Bamboo will use the default ElasticSearch settings, i.e., 'localhost'.

```python
>>> from bamboo import config, DataFrame

>>> config(hosts=['localhost'],
           sniff_on_start=True,
           sniff_on_connection_fail=True,
           sniffer_timeout=60)
>>> df = DataFrame(index='my_index')
```

#### Directly passing to dataframe

If different dataframes needs different credentials then a config object can be passed directly to the dataframe during initialization. Otherwise, it uses the global config defined above.

```python
>>> from bamboo.config import Config

>>> config_a = Config(hosts=['esnode1', 'esnode2'])
>>> config_b = Config(hosts=['esnode3', 'esnode4'])

>>> df_a = DataFrame(index='index_a', config=config_a)
>>> df_b = DataFrame(index='index_b', config=config_b)
```

#### Environmental variables

TBI

#### Config file

TBI

### DataFrame

The DataFrame is a represenation of a single elasticsearch index. Field names are loaded dyamically at initialization time. Several additional methods are provided for instrospecting the cluster.

Nested objects are referred to as namespaces within Bamboo.

```python
>>> from bamboo import DataFrame

>>> DataFrame.list_indices()
['my_index', 'index_a', 'index_b']

>>> df = DataFrame(index='my_index')

# Count of documents in index
>>> df.count()
46155585

# Get the data-types for a dataframe
>>> df.dtypes
{
    'name': 'string',
    'age': 'integer',
    'is_cat': 'boolean',
    'account': {
        'status': 'string',
        'balance': 'decimal',
        'expiration_date': 'date'
    }
}

# List fields
>>> df.fields
['name', 'age', 'is_cat']

# List namespaces
>>> df.namespaces
['account', 'other_namespace']

# List the fields within a namespace
>>> df.account.fields
['status', 'balance', 'expiration_date']
```

#### Field reference

Bamboo supports both dot and string-reference when specifying fields.

```python
>>> df.age
Integer(age)

>>> df['age']
Integer(age)

>>> df.account.status
String(status)

>>> df['account']['status']
String(status)

>>> df.get_namespace('account')['status']
String(status)

>>> status = df.account.status
>>> status.name
'account.status'
>>> status.parent
Namespace(account)
>>> status.dtype
'string'
```

### Querying

Generally, filtering works the same as you would use in Pandas.

Autocomplete works for field names. You can type period then tab (or whatever key mapping your IDE supports) and the available fields will be listed.

```python
>>> df = df[df.age > 21]
>>> df = df[df.is_cat == True]  # us `==`, not `is`
>>> df = df[df.account.status == 'good']
>>> df.take(10)
[{...}]
```

#### Boolean conditions

```python
>>> adult = df.age > 21
>>> cat = df.is_cat == True
>>> good_cat = df.account.status == 'good'

# and is represented by `&` and `+`
>>> df[adult + cat + good_cat]

# `&` is and. `|` is or
>>> df[adult | (cat & good_cat)]

# negation
>>> df[~adult & ~cat]

# conditions support comparing two dataframe objects
>>> df[adult] & df[cat]

# boosting the score of queries
>>> df[adult.boost(2.0) & cat.boost(1.5)]

# or use optional boost helper function
>>> from bamboo import boost
>>> df[boost(adult, 2.0) & boost(cat, 1.5)]
```

#### Filter

The filter parameter from ElasticSearch's bool query is made available as a DataFrame method. This filters out any documents that do not match any of the conditions sent to the filter parameter. However, matches to these conditions do not contribute to the overall score of a document.

```python
# matching a single condition
>>> df = df[df.is_cat==True]
>>> list(df.collect(include_score=True))
[{'_score': 1.0,
  'name': 'Purrrnest Hemingway',
  'is_cat': True}]

# filtering a single condition
>>> df = df.filter(df.is_cat==True)
>>> list(df.collect(include_score=True))
[{'_score': 0.0,
  'name': 'Purrrnest Hemingway',
  'is_cat': True}]

# filtering multiple conditions
>>> df = df.filter(df.is_cat==True, df.age=4)

# combining matching and filters
>>> df = df[df.is_cat==True]
>>> df = df.filter(df.age==4)

# filtering on boolean conditions
>>> adult = df.age > 21
>>> cat = df.is_cat == True
>>> df = df.filter(adult | cat)
```

#### Supported operators

| Operator  | Field Type    | Description              |
| :-------- | :-----------: | -----------------------: |
| ==        | all           | equal                    |
| !=        | all           | not equal                |
| >         | numeric, date | greater than             |
| >=        | numeric, date | greater than or equal to |
| <         | numeric, date | less than                |
| <=        | numeric, date | less than or equal to    |
| &         | all           | boolean and              |
| \|        | all           | boolean or               |
| ~         | all           | boolean negation         |

Method-based operators support autocomplete according to the field type. You can type period then tab (or whatever key mapping your IDE supports) and you will be provided with all the methods supported for a particular field.

One more method that is supported by all data types is `.isin`. You can use this to query that the value for a field falls within some set of values.

`is` as a syntactic operator is not supported. `==` should always be used instead. This includes when checking whether a field matches a boolean or a None.

##### Date operations

```python
>>> df[df.account.expiration_date >= '2019-11-01']

# automatic conversion to age (in days)
>>> df[df.account.activation_date.age < 30]
```

##### String operations

```python
# query where equal to a term
>>> df[df.name == 'cat']

# query where not equal to a term
>>> df[df.name != 'cat']

# query where a term contains a substring
>>> df[df.name.contains('freckles')]

# query where a term ends with a substring
>>> df[df.name.endswith('eckles')]

# query where a term startswith a substring
>>> df[df.name.startswith('freck')]

# query where a term matches a regular expression
>>> df[df.name.regexp(r'.*?\s+the\s+cat')]

# query where a term is a member of a set of values
>>> df[df.name.isin(['bengal', 'tabby', 'maine coon'])
```

#### Scripted queries (using painless)

```python
>>> source = "doc['age'].value < 10"
>>> script_condition = df.query(source)
>>> cat = df[df.is_cat == True]
>>> df[script_condition & cat]
```

### Document retrieval

#### Retrieve a single document

```python
# get a document according to document id
>>> df('document_id')
{...}

# only return certain fields
>>> df('document_id', fields=['name'])
{'name': 'Purrrnest Hemingway'}

# get a random document
>>> df.take(1)
[{...}]
```

#### Retrieve a collection of documents

```python
>>> df = df[df.is_cat == True]

# collect a limited number of documents
>>> df.take(5)
[{...}]  # array of five documents matching condition

# collect all matching documents. returns a generator
>>> df.collect()
<generator object __hits at 0x7fd6418fd0f0>

# collect has several optional parameters
>>> df.collect(fields=[],  # field names to include
               limit=10,  # number of documents to return
               preserve_order=False,  # sort documents by score
               include_score=False,  # include the score as a field
               include_id=False,  # include the document id as a field
               **es_kwargs)  # optional ElasticSearch params
<generator object __hits at 0x7fd6418fd0f0>

# apply limits to the number retrieved
>>> just_ten = df.limit(10)
>>> just_ten.collect()
<generator object __hits at 0x7fd6418fd0f0>

# convert query results to a pandas dataframe
>>> pd_df = df.to_pandas(fields=['age', 'name'])
>>> type(pd_df)
<class 'pandas.core.frame.DataFrame'>

# use ElasticSearch query language directly
>>> df.execute(body={'query': {'match_all': {}}})
<generator object __hits at 0x7fd6418fd0f0>
```

### Statistics and aggregations

#### All field types

```python
cat = df[df.is_cat == True]

# count of documents that satisfy condition
>>> cat.count()
42

# unique value counts for a field
>>> cat.name.value_counts(
        n=10,  # return top n counts
        normalize=False,  # return relative frequencies
        missing=None,  # how to treat missing documents
        **es_kwargs)  # optional ElasticSearch params
[('bengal', 10),
 ('tabby', 5),
 ('maine coon', 3)]

# count of distinct values
>>> cat.name.nunique(
        precision=3000,  # trade memory for accuracy
        **es_kwargs)  # optional ElasticSearch params
42
```

#### Numeric and date field types

```python
cat = df[df.is_cat == True]

# average
>>> cat.age.average(**es_kwargs)  # optional ElasticSearch params
10

# max
>>> cat.age.max(**es_kwargs)  # optional ElasticSearch params
22

# min
>>> cat.age.min(**es_kwargs)  # optional ElasticSearch params
1

# percentiles
>>> cat.age.percentiles(
        precision=100,  # trade memory for accuracy
        missing=None,  # how to treat missing documents
        **es_kwargs)  # optional ElasticSearch params
[{'key': 1.0, 'value': 0.0},
 {'key': 5.0, 'value': 0.0},
 {'key': 25.0, 'value': 0.003896170080258522},
 {'key': 50.0, 'value': 0.1366778221122553},
 {'key': 75.0, 'value': 2.7940936508189047},
 {'key': 95.0, 'value': 22.297436483702977},
 {'key': 99.0, 'value': 78.48965187864174}]

# summary statistics
>>> cat.age.describe(
        extended=False,  # return additional statistics
        missing=None,  # how to treat missing documents
        **es_kwargs)  # optional ElasticSearch params
{'avg': 10,
 'count': 42,
 'max': 22,
 'min': 1,
 'sum': 143}
```

#### Only numeric field types

```python
cat = df[df.is_cat == True]

# sum
>>> cat.age.sum(
        missing=None,  # how to treat missing documents
        **es_kwargs)  # optional ElasticSearch params
143

# median absolute deviation
>>> cat.age.median_absolute_deviation(
        precision=100,  # trade memory for accuracy
        missing=None,  # how to treat missing documents
        **es_kwargs)  # optional ElasticSearch params
3.056001642080668

# histogram
>>> cat.age.histogram(
        interval=50,  # bin size
        min_doc_count=1,  # minimum number of documents for a bin
        missing=None,  # how to treat missing documents
        **es_kwargs)  # optional ElasticSearch params
[('[0.0, 5.0)', 10),
 ('[5.0, 10.0)', 15),
 ('[10.0, 15.0)', 5),
 ('[15.0, 20)', 3)]

# percentile ranks
>>> cat.age.percentile_ranks(
        values=[5, 10, 15],  # desired percentage values
        precision=100,  # trade memory for accuracybin
        missing=None,  # how to treat missing documents
        **es_kwargs)  # optional ElasticSearch params
[{'key': 5.0, 'value': 10},
 {'key': 10.0, 'value': 15},
 {'key': 15.0, 'value': 5}]
```


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [Unreleased]

## [0.3.1]

### Changed

- Moved to Apache 2.0 license

### Added

- Option to install pandas

## [0.3.0]

### Added

- Filter parameter in bool query

### Fixed

- Raw query body is no longer symmetric, but instead follows the original order of conditions

## [0.2.5]

### Added

- MIT License.
- Flushed-out README.md
- Python3 compatibility

### Removed

- Upper limit on elasticsearch version.

## [0.2.4]

### Fixed

- Bug where aggregating a datetime field when there are no matching documents.

## [0.2.3]

### Fixed

- Bug in certain nested conditions.

## [0.2.2]

### Added

- HTML repr in notebook.

### Fixed

- Bug where conditions would leak between bool query types.
- Bug where an inversion in root order of operations could result in different query.

## [0.2.1]

### Fixed

- NameError in Field.isin

## [0.2.0] - 2019-09-16

### Added

- DataFrame as replacement for ElasticDataFrame.
- Boost method for queries.
- Boost helper function to wrap queries.
- Allow returning score in results.
- Allow returning id in results.
- Field.isin method for searching multiple terms.
- Query objects.
- Allow passing distinct config objects to dataframe in addition to setting a single global object used by all dataframes.
- DataFrame.query method for applying custom script queries.

### Changed

- Deprecated ElasticDataFrame.
- Moved query classes from utils.py to query.py
- Changed DataFrame.take from generator to return a list.
- Exposed DataFrame.execute as a public method.

### Removed

- Raw results option from DataFrame.collect.

### Fixed

- Bug where DataFrame.collect with `limit` set and `raw=True` would not return results.
- Bug where deeply nested queries returned incorrect results.


bamboo/fields.py
# This Source Code Form is subject to the terms of the Mozilla Public
# License, v. 2.0. If a copy of the MPL was not distributed with this file,
# You can obtain one at http://mozilla.org/MPL/2.0/.
"""Field and namespace objects representing Elasticsearch field types."""
import warnings
from abc import ABCMeta
from copy import deepcopy
from datetime import datetime, timedelta
from functools import wraps

from . import queries


def check_inversion(func):
    """Decorate a method for invertible operations."""
    @wraps(func)
    def wrapper(obj, *args, **kwargs):
        condition = func(obj, *args, **kwargs)
        if obj._inverted is True:
            condition = queries.Bool(must_not=condition)
        return condition
    return wrapper


class Base(object):
    """Parent class for field and namespace classes.

    Attributes:
        name: Full field name including any parents
        parent: Parent for a field
    """

    __metaclass__ = ABCMeta

    def __init__(self, name, parent):
        """Init Base.

        Args:
            name (str): The name for the class
        """
        self.parent = parent
        self._name = name
        self._inverted = False

    def __invert__(self):
        new = deepcopy(self)
        new._inverted = True
        return new

    def __repr__(self):
        return '{}({})'.format(type(self).__name__, self._name)

    @property
    def name(self):
        """Full field name including any parents."""
        if isinstance(self.parent, Base):
            return self.parent.name + '.' + self._name
        return self._name

    @property
    def root(self):
        """Dataframe where a field is attached."""
        return self.__get_root(self)

    @staticmethod
    def __get_root(obj):
        """Recurse parent objects until root dataframe."""
        parent = obj.parent
        if not isinstance(parent, Base):
            return parent
        return Base.__get_root(parent)

    @check_inversion
    def exists(self):
        """Condition checking whether the field exists."""
        return queries.Exists(self.name)


class Namespace(Base):
    """Namespace field meant to hold children fields.

    Attributes:
        name: Full field name including any parents
        parent: Parent for a field
        fields: Children fields held by namespace
    """

    def __getitem__(self, key):
        return getattr(self, key)

    @property
    def fields(self):
        """Fields that exist within a namespace."""
        return [i._name
                for i in vars(self).values()
                if isinstance(i, Field)]


class Field(Base):
    """Base field class for different dtypes.

    Attributes:
        name: Full field name including any parents
        parent: Parent for a field
    """

    __metaclass__ = ABCMeta

    @check_inversion
    def __eq__(self, value):
        return queries.Term(self.name, value)

    def __ne__(self, value):
        condition = queries.Term(self.name, value)
        if self._inverted:
            return condition
        return queries.Bool(must_not=condition)

    @check_inversion
    def isin(self, values):
        """Condition checking whether multiple terms are in a field's value.

        Args:
            values (list): Sequence of values to check
        """
        return queries.Terms(self.name, values)

    def value_counts(self, n=10, normalize=False, missing=None, **es_kwargs):
        """Get the unique value counts for a field.

        The results are approximate if the number of unique terms is
        greater than the `n` parameter. The higher the requested `n` is,
        the more accurate the results will be, but also, the more expensive
        it will be to compute the final results.

        Args:
            n (int, optional): The number of value counts to return.
                Defaults to 10.
            normalize (bool, optional): If True then the object returned
                will contain the relative frequencies of the unique values.
                Defaults to False.
            missing (num, optional): How documents that are missing a
                value should be treated. Defaults to ignore.

        Returns:
            List[tuple]: Tuple consiting of (key, count) ordered by descending count.
                `OTHER` is included in counts if the `n` parameter is not
                enough to cover all counts.
        """
        results = self._simple_aggregation('terms',
                                           missing=missing,
                                           buckets=False,
                                           size=n,
                                           es_kwargs=es_kwargs)
        counts = [(i['key'], i['doc_count']) for i in results['buckets']]
        other_counts = results['sum_other_doc_count']
        error_bound = results['doc_count_error_upper_bound']
        if counts and error_bound > counts[-1][1]:
            msg = """
                The current `n` parameter may result in an approximation
                where a term is not included in the returned results ({})
            """.format(error_bound)
            warnings.warn(msg)
        if other_counts:
            counts.append(('OTHER', other_counts))
        if normalize:
            total = sum(i[1] for i in counts)
            counts = [(i[0], i[1] / float(total)) for i in counts]
        return counts

    def nunique(self, precision=3000, **es_kwargs):
        """Get the approximate count of distinct values.

        Args:
            precision (int, optional): Define a unique count below which
                counts are expected to be close to accurate. Trade memory
                for accuracy. Above this value, counts might become a bit
                more fuzzy. The maximum supported value is 40000, thresholds
                above this number will have the same effect as a threshold
                of 40000. Defaults to 3000.

        Returns:
            int: Distinct count
        """
        result = self._simple_aggregation('cardinality',
                                          precision_threshold=precision,
                                          es_kwargs=es_kwargs)
        return result['value']

    def _simple_aggregation(self, key, buckets=True, **params):
        """Execute simple aggregation query with no query hits."""
        if self.root._limit:
            warnings.warn("Limits are not applied in aggregations.")
        es_kwargs = params.pop('es_kwargs')
        params = {k: v for k, v in params.items() if v is not None}
        body = dict(self.root._body, **{
            'aggs': {
                key: {
                    key: dict({'field': self.name}, **params)
                }
            }
        })
        results = self.root.execute(body, size=0, **es_kwargs)
        results = results['aggregations'][key]
        if buckets:
            return results.get('buckets', results)
        return results


class AggregationMixin(object):
    """Adds aggregations to a field."""

    __metaclass__ = ABCMeta

    def average(self, **es_kwargs):
        """Get the average value for a field.

        Returns:
            float: Average
        """
        result = self._simple_aggregation('avg', es_kwargs=es_kwargs)
        return result['value']

    def max(self, **es_kwargs):
        """Get the max value for a field.

        Returns:
            float: Max
        """
        result = self._simple_aggregation('max', es_kwargs=es_kwargs)
        return result['value']

    def min(self, **es_kwargs):
        """Get the min value for a field.

        Returns:
            float: Max
        """
        result = self._simple_aggregation('min', es_kwargs=es_kwargs)
        return result['value']

    def percentiles(self, missing=None, precision=100, **es_kwargs):
        """Get percentiles for a filed.

        Percentiles are calculated over the range of [ 1, 5, 25, 50, 75, 95, 99 ]

        Args:
            missing (num, optional): How documents that are missing a
                value should be treated. Defaults to ignore.
            precision (int, optional): Defines the accuracy - memory tradeoff.
                Larger precision result in greater accuracy at the cost of
                greater memory usage and computation time. Defaults to 1000.

        Notes:
            - Accuracy is proportional to q(1-q). This means that extreme
                percentiles (e.g. 99%) are more accurate than less extreme
                percentiles, such as the median.
            - For small sets of values, percentiles are highly accurate
                (and potentially 100% accurate if the data is small enough).
            - As the quantity of values in a bucket grows, the algorithm
                begins to approximate the percentiles. It is effectively
                trading accuracy for memory savings.

        Returns:
            List[dict]: List of dicts with `key` as the perecentile range
                and `value` as the point where that percentile occurs.
        """
        result = self._simple_aggregation('percentiles',
                                          keyed=False,
                                          tdigest={'compression': precision},
                                          missing=missing,
                                          es_kwargs=es_kwargs)
        return result['values']

    def describe(self, extended=False, missing=None, **es_kwargs):
        """Get statistical details for a field.

        Args:
            extended (bool, optional): Increase the number of statistics returned.
                Defaults to False.
            missing (num, optional): How documents that are missing a
                value should be treated. Defaults to ignore.

        Returns:
            dict: Always returns `count`, `min`, `max`, `average`, `sum`.
                If `extended` is True then also returns `sum_of_squares`,
                `variance`, `std_deviation`, and `std_deviation_bounds`.

        """
        key = 'extended_stats' if extended else 'stats'
        return self._simple_aggregation(key, missing=missing, es_kwargs=es_kwargs)

    def histogram(self, interval=50, min_doc_count=1, missing=None, **es_kwargs):
        """Get a count of values for a field bucketed by interval.

        Args:
            interval (number, optional): Bin size for counts. Defaults to 50.
            min_doc_count (int, optional): Minimum number of documents for
                a bin to show up.
            missing (num, optional): How documents that are missing a
                value should be treated. Defaults to ignore.

        Returns:
            List[tuple]: Tuple consiting of (range, count).
        """
        def make_key(key):
            return '[{}, {})'.format(key, key+interval)
        results = self._simple_aggregation('histogram',
                                           interval=interval,
                                           min_doc_count=min_doc_count,
                                           missing=missing,
                                           es_kwargs=es_kwargs)
        return [(make_key(i['key']), i['doc_count']) for i in results]


class RangeMixin(object):
    """Adds comparator operators to a Field dtype."""

    __metaclass__ = ABCMeta

    @check_inversion
    def __lt__(self, value):
        return queries.Range(self.name).less_than(value)

    @check_inversion
    def __le__(self, value):
        return queries.Range(self.name).less_than_or_equal(value)

    @check_inversion
    def __gt__(self, value):
        return queries.Range(self.name).greater_than(value)

    @check_inversion
    def __ge__(self, value):
        return queries.Range(self.name).greater_than_or_equal(value)


class Numeric(Field, RangeMixin, AggregationMixin):
    """Numeric base for conditions and aggregations."""

    __metaclass__ = ABCMeta

    def percentile_ranks(self, values, missing=None, precision=100, **es_kwargs):
        """Get the percentage of observed values which are below a certain value.

        Args:
            values (List[int]): The values for which percentages are desired
            missing (num, optional): How documents that are missing a
                value should be treated. Defaults to ignore.
            precision (int, optional): Defines the accuracy - memory tradeoff.
                Larger precision result in greater accuracy at the cost of
                greater memory usage and computation time. Defaults to 1000.

        Notes:
            - Accuracy is proportional to q(1-q). This means that extreme
                percentiles (e.g. 99%) are more accurate than less extreme
                percentiles, such as the median.
            - For small sets of values, percentiles are highly accurate
                (and potentially 100% accurate if the data is small enough).
            - As the quantity of values in a bucket grows, the algorithm
                begins to approximate the percentiles. It is effectively
                trading accuracy for memory savings.

        Returns:
            List[dict]: List of dicts with `key` as the value and `value` as the
                percentage of items falling below that value.
        """
        result = self._simple_aggregation('percentile_ranks',
                                          keyed=False,
                                          values=values,
                                          tdigest={'compression': precision},
                                          missing=missing,
                                          es_kwargs=es_kwargs)
        return result['values']

    def sum(self, missing=None, **es_kwargs):
        """Get the sum of observed values.

        Args:
            missing (num, optional): How documents that are missing a
                value should be treated. Defaults to ignore.

        Returns:
            float: The sum
        """
        result = self._simple_aggregation('sum', missing=missing, es_kwargs=es_kwargs)
        return result['value']

    def median_absolute_deviation(self, missing=None, precision=1000, **es_kwargs):
        """Approximates the median absolute deviation for a field.

        Median absolute deviation is a measure of variability. It is a
        robust statistic, meaning that it is useful for describing data
        that may have outliers, or may not be normally distributed. For
        such data it can be more descriptive than standard deviation.

        https://en.wikipedia.org/wiki/Median_absolute_deviation

        Args:
            missing (num, optional): How documents that are missing a
                value should be treated. Defaults to ignore.
            precision (int, optional): Defines the accuracy - memory tradeoff.
                Larger precision result in greater accuracy at the cost of
                greater memory usage and computation time. Defaults to 1000.

        Returns:
            float: Median absolute deviation
        """
        result = self._simple_aggregation('median_absolute_deviation',
                                          compression=precision,
                                          missing=missing,
                                          es_kwargs=es_kwargs)
        return result['value']


class Integer(Numeric):
    """Integer field dtype."""

    dtype = 'integer'


class Float(Numeric):
    """Float field dtype."""

    dtype = 'float'


class Decimal(Numeric):
    """Decimal field dtype."""

    dtype = 'decimal'


class Boolean(Field):
    """Boolean field dtype."""

    dtype = 'boolean'


class String(Field):
    """String field dtype."""

    dtype = 'string'

    def match(self, term):
        """Condition checking whether term occurs in an analyzed field.

        Args:
            term (str): The term being matched
        """
        return queries.Match(self.name, term)

    @check_inversion
    def regexp(self, value):
        """Condition checking whether regular expression is to a field's value.

        Args:
            value (str): Regular expression
        """
        return queries.Regexp(self.name, value)

    @check_inversion
    def contains(self, value):
        """Condition checking whether string is contained in a field's value.

        This naive query is going to be significantly slower than using a
        regexp query or a match query with analyzed fields.

        Args:
            value (str): String that should exist
        """
        return queries.Wildcard(self.name, '*{}*'.format(value))

    @check_inversion
    def startswith(self, value):
        """Condition checking whether a field's value starts with a string.

        Args:
            value (str): String that a field's value should start with
        """
        return queries.Prefix(self.name, value)

    @check_inversion
    def endswith(self, value):
        """Condition checking whether a field's value ends with a string.

        This naive query is going to be significantly slower than using a
        regexp query or a match query with analyzed fields.

        Args:
            value (str): String that a field's value should end with
        """
        return queries.Wildcard(self.name, '*{}'.format(value))


class Date(Field, RangeMixin, AggregationMixin):
    """Date/datetime field dtype."""

    dtype = 'date'

    class Age(Field, RangeMixin):
        """Delta from datetime field dtype."""

        dtype = 'age'

        def age_to_dt(func):
            """Decorate a method to supply delta from today as date."""
            @wraps(func)
            def wrapper(obj, value):
                delta = datetime.today() - timedelta(value)
                return func(obj, delta)
            return wrapper

        __eq__ = age_to_dt(Field.__eq__)
        __ne__ = age_to_dt(Field.__ne__)
        __lt__ = age_to_dt(RangeMixin.__gt__)
        __le__ = age_to_dt(RangeMixin.__ge__)
        __gt__ = age_to_dt(RangeMixin.__lt__)
        __ge__ = age_to_dt(RangeMixin.__le__)

    @property
    def age(self):
        """Condition treating the value being compared as delta since today."""
        return self.Age(self.name, None)

    def _epoch_to_dt(func):
        @wraps(func)
        def wrapper(obj, *args, **kwargs):
            result = func(obj, *args, **kwargs)
            if result:
                return datetime.fromtimestamp(result / 1000)
        return wrapper

    average = _epoch_to_dt(AggregationMixin.average)
    max = _epoch_to_dt(AggregationMixin.max)
    min = _epoch_to_dt(AggregationMixin.min)

    def histogram(self, interval=50, min_doc_count=1):
        """Get a count of values for a field bucketed by interval.

        Args:
            interval (number, optional): Bin size for counts. Defaults to 50.
            min_doc_count (int, optional):
        """
        raise NotImplementedError("Bug Aaron to implement it.")


class Dummy(Field):
    """Dummy field for not implemented dtypes."""

    dtype = 'dummy'
